### PR TITLE
impl Stream and remove the "eager retry" logic

### PR DESCRIPTION
**Pull request by loyd**
* Added a `Stream` instance.
* Replaced deprecated lints and methods.
* Removed repolling on `Pending`. It's a controversial change, but I don't understand why the lib contains that logic. Firstly, it's a pretty rare situation, but it adds extra CAS on every `Pending`. Secondly, it doesn't seem to be the library's responsibility (and It breaks our tests (WS pool with tricky logic)).

### CODE REVIEW

File: src/lib.rs
Comment (jonhoo) [inline src/lib.rs:136]: Why change the capitalization of these?

File: src/lib.rs
Comment (jonhoo) [inline src/lib.rs]: This can be `Relaxed` on failure I'm fairly certain.

File: Cargo.toml
Comment (jonhoo) [inline Cargo.toml]: I'd like to see a feature to specifically opt into this for those who may wish to avoid a dependency on `futures-core`.

File: Cargo.toml
Comment (jonhoo) [inline Cargo.toml:19]: I wish we could get rid of this in favor of using `std::task::Wake`, which was stabilized in 1.51. Unfortunately, `std` doesn't contain an equivalent for `futures_task::waker_ref`, which would mean we either need to clone the `Arc` on every poll or keep two `Arc`s around in the wrapper all the time.

File: src/lib.rs
Comment (jonhoo) [inline src/lib.rs:181]: I agree we can remove the "eager retry" logic. It may improve performance marginally in some cases, but it's also not really our responsibility to mitigate such races. I do think we should include a brief comment explaining why a race won't result in a lost wake-up though. (It's because whatever may have raced with us must have also called `.wake()`).

**Review by jonhoo — CHANGES_REQUESTED** [review]
These are good changes, thank you! Left some notes inline.

Also, would you mind rebasing against `master` to pick up the CI changes I just made?

File: src/lib.rs
Comment (loyd) [reply src/lib.rs]: Agree, fixed

File: Cargo.toml
Comment (loyd) [reply Cargo.toml]: * added a `stream` feature, disabled by default
* noted about it in the crate-level comment
* enabled `doc_auto_cfg` for `docs.rs` to show a badge next to a `Stream` instance
* fixed a test to actually use `Stream::poll_next`
* added a trivial test to check that `size_hint` is propagated, mostly for coverage

File: src/lib.rs
Comment (loyd) [reply src/lib.rs:136]: It's mostly a subjective question. However, I have several arguments:
* In [rustc](https://github.com/search?p=1&q=%22safety%3A%22+repo%3Arust-lang%2Frust+extension%3Ars+language%3ARust&type=Code) and [rust-analyzer](https://github.com/rust-lang/rust-analyzer/search?q=safety):
  ```sh
  forks/rust [master] » rg '// safety:' | wc -l
  5
  forks/rust [master] » rg '// Safety:' | wc -l
  211
  forks/rust [master] » rg '// SAFETY:' | wc -l
  1312
  ```
* Some editors (e.g. vim) [highlights](https://github.com/sheerun/vim-polyglot/blob/730dcb02caab60a6ae5d8b4bdc16d290041061ec/syntax/rust.vim#L236) `SAFETY`, but doesn't highlight `safety`.

If you insist, I'll change it back, of course.

File: src/lib.rs
Comment (loyd) [reply src/lib.rs]: Maybe we should use `loom` to have more guarantees here. It seems that `SeqCst` is excessive in any case.
But it's a subject of another PR.

File: Cargo.toml
Comment (loyd) [reply Cargo.toml:19]: I would like to avoid extra `fetch_add` at the hot path, although it violates a cache coherence only if the waker is shared between cores; ~~that's a rare case~~.
Also, I'm not sure that fully understand the idea of your second proposal (with two `Arc`s)

Anyway, it goes beyond this PR.

File: src/lib.rs
Comment (loyd) [reply src/lib.rs:181]: Added. Is it enough?

File: src/lib.rs
Comment (jonhoo) [reply src/lib.rs:136]: Nope, seems fine :+1:

File: src/lib.rs
Comment (jonhoo) [reply src/lib.rs]: Yep, I'd love to have loom tests for this crate! Would be happy to take a look at a separate PR for that if you're up for it :+1:

File: Cargo.toml
Comment (jonhoo) [reply Cargo.toml:19]: The "two Arc" solution would be to keep both the `Arc` we have right now which holds a concrete type that we can use to get at inner fields but cannot easily be used as a `Waker`, _and_ a `Waker::from(Arc::clone(&waker))` which can be passed in as the `Waker` to `poll` but doesn't have access to private fields.

But yes, for a different PR.

File: src/lib.rs
Comment (jonhoo) [reply src/lib.rs:181]: I want to tweak it a little but can do that manually after merging!

**Review by jonhoo — APPROVED** [review]
Looks good, thanks!

File: src/lib.rs
Comment (jonhoo) [inline src/lib.rs]:
```suggestion
        // `poll_fn` can call `wake()` immediately and, hence, `awoken` can be changed here.
        // However, by the `Future::poll` contract, it must have called `.wake`, which means
        // we've called `wake` on _our_ waker, which means we'll be polled again later.
```

**Comment by jonhoo** [comment]
Looks like two CI failures. One is because the rustdoc lint hadn't yet been renamed in 1.51, which we still support, so [its build fails](https://github.com/jonhoo/strawpoll/actions/runs/2997523283/jobs/4967495665). The other is because, seemingly, something in our dependency graph [isn't truthfully representing](https://github.com/jonhoo/strawpoll/actions/runs/2997523285/jobs/4967495710) their minimum required dependency versions. You may have to explicitly list `async-stream-impl` in our `Cargo.toml` with a higher version requirement (and perhaps send a PR to `tokio-stream` to bump its listed dependency version?).

**Comment by loyd** [comment]
@jonhoo, could you run lints? I want to fix them finally =)

**Comment by loyd** [comment]
https://github.com/tokio-rs/tokio/pull/5347

File: Cargo.toml
Comment (jonhoo) [inline Cargo.toml]: That PR is now merged! :tada:

**Comment by jonhoo** [comment]
Happy to merge this once we remove that now-unneeded line from Cargo.toml :tada: 
I'll do a release shortly after.

**Comment by loyd** [comment]
So, are you waiting for the new release of `tokio-test`? The latest release was in May, this is not a quick process, sadly. I think specifying `async-stream` is good enough to make a release. Maybe the comment should be `remove once tokio-test v0.4.3 is released` =)

**Comment by jonhoo** [comment]
Ah, I thought they did a release after merging the PR. In that case just updating the comment is fine :+1:

**Review by jonhoo — APPROVED** [review]

**Comment by jonhoo** [comment]
Released in 0.2.3 :tada: Thanks for sticking with it :)

**Pull request merged**
